discovery-infra/start_discovery.py

```diff
@@ -27,61 +29,110 @@ def _create_ip_address_list(node_count, starting_ip_addr):
 
 
 # Filling tfvars json files with terraform needed variables to spawn vms
-def fill_tfvars(image_path, storage_path, master_count, nodes_details):
-    if not os.path.exists(consts.TFVARS_JSON_FILE):
-        Path(consts.TF_FOLDER).mkdir(parents=True, exist_ok=True)
-        copy_tree(consts.TF_TEMPLATE, consts.TF_FOLDER)
-
-    with open(consts.TFVARS_JSON_FILE) as _file:
+def fill_tfvars(
+        image_path,
+        storage_path,
+        master_count,
+        nodes_details,
+        tf_folder
+        ):
+    tfvars_json_file = os.path.join(tf_folder, consts.TFVARS_JSON_NAME)
+    with open(tfvars_json_file) as _file:
         tfvars = json.load(_file)
-    master_starting_ip = str(
-        ipaddress.ip_address(
-            ipaddress.IPv4Network(nodes_details["machine_cidr"]).network_address
-        )
-        + 10
-    )
-    worker_starting_ip = str(
-        ipaddress.ip_address(
-            ipaddress.IPv4Network(nodes_details["machine_cidr"]).network_address
-        )
-        + 10
-        + int(tfvars["master_count"])
+
+    master_count = min(master_count, consts.NUMBER_OF_MASTERS)
+    nodes_ips = _find_free_ips(
+        machine_cidr=nodes_details['machine_cidr'],
+        start_ip=IPNetwork(nodes_details['machine_cidr']).ip + 10,
+        num_required_ips=master_count + nodes_details['worker_count']
     )
+
     tfvars["image_path"] = image_path
-    tfvars["master_count"] = min(master_count, consts.NUMBER_OF_MASTERS)
-    tfvars["libvirt_master_ips"] = _create_ip_address_list(
-        min(master_count, consts.NUMBER_OF_MASTERS), starting_ip_addr=master_starting_ip
-    )
+    tfvars["master_count"] = master_count
+    tfvars['libvirt_master_ips'] = nodes_ips[:master_count]
+    tfvars['libvirt_worker_ips'] = nodes_ips[master_count:]
     tfvars["api_vip"] = _get_vips_ips()[0]
-    tfvars["libvirt_worker_ips"] = _create_ip_address_list(
-        nodes_details["worker_count"], starting_ip_addr=worker_starting_ip
-    )
     tfvars["libvirt_storage_pool_path"] = storage_path
     tfvars.update(nodes_details)
 
-    with open(consts.TFVARS_JSON_FILE, "w") as _file:
+    with open(tfvars_json_file, "w") as _file:
         json.dump(tfvars, _file)
 
 
+def _find_free_ips(machine_cidr, start_ip, num_required_ips):
```

Comment:
why we need to find _find_free_ips in subnet? we can't have 2 clusters in the same subnet so why it is interesting?

Reply:
u need to validate subnet is free and not used and not ips

Reply:
even worth if one of the ips is not free it will be a mistake to try to create cluster u will not know what happened to your cluster on failure in this case.